README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# CHOIRBM

<!-- badges: start -->
<!-- badges: end -->

CHOIRBM is a collection of plotting utilities for the [Collaborative
Health Outcomes Information Registry’s](https://choir.stanford.edu) Body
Map (CBM). The CBM is an instrument for assessing the distribution of a
patient’s pain, and has been validated in a paper published in [*Pain
Reports* (Scherrer et al
2021)](https://doi.org/10.1097/pr9.0000000000000880). The package is
built on top of the popular R plotting package `ggplot2`, and returns
plots as ggplot objects.

Sample of the CBM in clinical use:
<img src="man/figures/cbm-journal.jpeg" width="100%" />

## Installation

~~You can install the released version of CHOIRBM from
[CRAN](https://CRAN.R-project.org) with:~~ (Not on CRAN…yet)

``` r
install.packages("CHOIRBM")
```

Or install from GitHub with:

``` r
devtools::install_github("emcramer/CHOIRBM")
remotes::install_github("emcramer/CHOIRBM")
```

## Examples

### Plotting the Male CBM

This is a basic example which shows you how to plot the front and back
parts of the male CHOIR Body Map:

``` r
library(CHOIRBM)

# generate some random example data
set.seed(123)
ids <- as.character(c(seq.int(101, 136, 1), seq.int(201, 238, 1)))
values <- data.frame(
  id = ids
  , value = runif(length(ids))
  , ucolors = rainbow(length(ids))
  , group = ifelse(as.numeric(ids) < 200, "Front", "Back")
)

# plot the data on the front of the CHOIR body map
plot_male_choirbm(values, "value")
```

<img src="man/figures/README-male_example-1.png" width="100%" />

``` r
# plot each segment of the bodymap as a different color
plot_male_choirbm(values, "ucolors")
```

<img src="man/figures/README-male_example-2.png" width="100%" />

### Plotting the female CBM

And an additional example with the female CHOIR Body Map:

``` r
library(CHOIRBM)

# generate some random example data
exdata <- gen_example_data()

# plot the data on the front of the CHOIR body map
plot_female_choirbm(exdata, "value")
```

<img src="man/figures/README-female_example-1.png" width="100%" />

Check the vignettes for more examples such as plotting individual
patients, highlighting specific segments, and changing color schemes.

## Citations:

> Scherrer, Kristen Hymel; Ziadni, Maisa S; Kong, Jiang-Tia; Sturgeon,
> John A; Salmasi, Vafia; Hong, Juliette; Cramer, Eric; Chen, Abby L;
> Pacht, Teresa; Olson, Garrick; Darnall, Beth D; Kao, Ming-Chih;
> Mackey, Sean. Development and validation of the Collaborative Health
> Outcomes Information Registry body map, PAIN Reports: January/February
> 2021 - Volume 6 - Issue 1 - p e880 doi:
> [10.1097/PR9.0000000000000880](https://doi.org/10.1097/pr9.0000000000000880)


NEWS.md
# CHOIRBM 0.0.2

* Fixed a url in the README which was broken.

# CHOIRBM 0.0.1

* Simple functionality for plotting male and female CHOIR Body Maps (CBMs).
  - Plotting functions for male and female body maps.
  - Utility function for converting a delimited body map into a plotting-ready
    data frame.
  - Utility function for generating a random data set.
  - Sample data set for examples of data processing and plotting.

---

# CHOIRBM 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(CHOIRBM)

test_check("CHOIRBM")


tests/testthat/test-string-to-map.R
context("String maps equal")
test_that("string_to_map yields valid map", {

  # create a simple body map data frame
  ids <- as.character(c(seq.int(101, 136, 1), seq.int(201, 238, 1)))
  sample_df <- data.frame(
    id = ids
    , value = c(1, 1, rep(0, 34), 1, 1, rep(0, 36))
    , group = ifelse(as.numeric(ids) < 200, "Front", "Back")
  )

  # string representation of the data frame
  cbm_str <- "101,102,201,202"

  # generate body map df from string
  test_df <- string_to_map(cbm_str)

  # test each column for equality
  expect_equal(test_df[["id"]], sample_df[["id"]])
  expect_equal(test_df[["value"]], sample_df[["value"]])
  expect_equal(test_df[["group"]], sample_df[["group"]])
})


R/validation.R
#' CHOIR Body Map data for approximately 7,000 patients
#'
#' A non-identifiable, simulated data set generated by randomly permuting
#' data from the CHOIR Body Map validation study.
#'
#' @docType data
#'
#' @usage data(validation)
#'
#' @format An object of class \code{"data.frame"}
#' \describe{
#'  \item{id}{A randomly generated numeric code for each patient.}
#'  \item{gender}{The patient's gender.}
#'  \item{race}{The patient's race.}
#'  \item{age}{The patient's age.}
#'  \item{bodymap_regions_csv}{The patient's CHOIR Body Map in a comma
#'  separated string.}
#'  \item{score}{A simulated pain score for demonstration purposes.}
#' }
#'
#' @references This data set was derived from the data collected during the
#' study validating the CHOIR Body Map as an instrument for recording a
#' patient's anatomical pain location.
#' \doi{10.1097/pr9.0000000000000880}
#'
#' @keywords datasets
#' @examples
#'
#' data(validation)
#' head(validation)
#'
"validation"


R/gen_example_data.R
#' Generate Simple Example Data
#'
#' Creates a data frame with CHOIR Body Map segment IDs and a randomly
#' associated value. Also adds grouping information for facetting while
#' plotting.
#'
#' @param seed integer to seed the random number generator
#' @return values data.frame
#' @export
#'
#' @examples
#' cbm_df <- gen_example_data()
#' head(cbm_df)
#'
#' @importFrom stats runif
#'
gen_example_data <- function(seed = 123) {
  ids <- as.character(c(seq.int(101, 136, 1), seq.int(201, 238, 1)))
  set.seed(seed = seed)
  values <- data.frame(
    id = ids
    , value = runif(length(ids))
    , group = factor(
      ifelse(
        as.numeric(ids) < 200
        , "Front"
        , "Back"
        )
      , levels = c("Front", "Back")
      )
  )
  return(values)
}


R/plot_male_choirbm.R
#' Plot the male CHOIR Body Map
#'
#' Creates a new plot of the male CHOIR body map.
#'
#' @param df data.frame
#' @param value string
#' @return ggrob
#' @export
#'
#' @examples
#' cbm_df <- gen_example_data()
#' plot_male_choirbm(cbm_df, "value")
#'
#' @importFrom ggplot2 ggplot aes geom_polygon scale_y_reverse facet_wrap
#' @importFrom ggplot2 theme_void theme
#' @importFrom rlang .data
#'
plot_male_choirbm <- function(df, value) {
  positions <- rbind(male_front_poly_coords_df, male_back_poly_coords_df)
  datapoly <- merge(df, positions, by = "id")
  p <- ggplot(datapoly) +
    aes(x = x, y = y) +
    geom_polygon(
      aes(
        fill = .data[[value]]
        , group = .data[["id"]]
        )
      , color = "black", size = 1.5
      ) +
    scale_y_reverse() +
    facet_wrap(~.data[["group"]]) +
    theme_void() +
    theme(legend.position = "None")
  return(p)
}


R/plot_female_choirbm.R
#' Plot the male CHOIR Body Map
#'
#' Creates a new plot of the front and back of the female CHOIR body map.
#'
#' @param df data.frame
#' @param value string
#' @return ggrob
#' @export
#'
#' @examples
#' cbm_df <- gen_example_data()
#' plot_female_choirbm(cbm_df, "value")
#'
#' @importFrom ggplot2 ggplot aes geom_polygon scale_y_reverse facet_wrap
#' @importFrom ggplot2 theme_void theme
#' @importFrom rlang .data
#'
plot_female_choirbm <- function(df, value) {
  positions <- rbind(female_front_poly_coords_df, female_back_poly_coords_df)
  datapoly <- merge(df, positions, by = "id")
  p <- ggplot(datapoly) +
    aes(x = x, y = y) +
    geom_polygon(
      aes(
        fill = .data[[value]]
        , group = .data[["id"]]
        )
      , color = "black"
      , size = 1.5
      ) +
    scale_y_reverse() +
    facet_wrap(~.data[["group"]]) +
    theme_void() +
    theme(legend.position = "None")
  return(p)
}


R/string_to_map.R
#' Converts a comma-separated string to a CHOIR BM
#'
#' Takes a string of IDs that are separated by a comma and converts the
#' information into a data frame with a binary indication of whether or not
#' an ID appeared. Useful for plotting an individual's CHOIR BM or for
#' isolating particular sections to highlight.
#'
#' @param map_str The delimited CBM string.
#' @param delim The delimiter for the CBM string.
#'
#' @return ret_df data.frame with all of the CHOIR BM segment IDs with a 1
#' if the segment was present and 0 otherwise.
#' @export
#'
#' @examples
#' cbm_str <- "101,102,103,104,201,202"
#' cbm_df <- string_to_map(cbm_str)
#' #plot in a male or female bodymap...
#' plot_male_choirbm(cbm_df, "value")
#'
#' @importFrom stringr str_split
#'
string_to_map <- function(map_str = "", delim = ",") {
  ids <- as.character(c(seq.int(101, 136, 1), seq.int(201, 238, 1)))
  ret_df <- data.frame(
    id = ids
    , group = ifelse(as.numeric(ids) < 200, "Front", "Back")
    , value = as.integer(ids %in% unlist(str_split(map_str, delim)))
  )
  return(ret_df)
}
